File: 06-state-management/golang/gateway.go
package main

import (
	"encoding/json"
	"fmt"
	"os"
	"time"
)

// ── ACP types ─────────────────────────────────────────────────────────────────

// ACPMessage is the unified in-process message from any channel to the Agent.
type ACPMessage struct {
	ID        string
	SessionID string
	Channel   string
	Content   string
	Timestamp int64
}

// AgentReply carries one streaming token ("delta"), the final text ("reply"),
// history on reconnect ("history"), or an error ("error") back to the channel.
type AgentReply struct {
	Type      string // "delta" | "reply" | "history" | "error"
	ID        string
	SessionID string
	Channel   string
	Content   string
}

// ── Gateway ───────────────────────────────────────────────────────────────────

type Gateway struct {
	adapters map[string]ChannelAdapter
	agent    *Agent
	db       *DB
}

func newGateway(agent *Agent, db *DB) *Gateway {
	return &Gateway{adapters: make(map[string]ChannelAdapter), agent: agent, db: db}
}

func (g *Gateway) register(adapter ChannelAdapter) {
	g.adapters[adapter.name()] = adapter
	adapter.onMessage(func(raw ACPMessage) {
		go g.dispatch(raw)
	})
	// Wire up history loader for web adapter so reconnecting browsers get history.
	if wa, ok := adapter.(*webAdapter); ok {
		wa.historyLoader = func(sessionID string) []HistoryEntry {
			if g.db == nil {
				return nil
			}
			history, _ := g.db.loadHistory(sessionID)
			return history
		}
	}
}

func (g *Gateway) dispatch(raw ACPMessage) {
	msg := ACPMessage{
		ID:        raw.ID,
		SessionID: resolveSessionID(raw.Channel, raw.SessionID),
		Channel:   raw.Channel,
		Content:   raw.Content,
		Timestamp: raw.Timestamp,
	}

	adapter, ok := g.adapters[msg.Channel]
	if !ok {
		fmt.Fprintf(os.Stderr, "[gateway] unknown channel: %s\n", msg.Channel)
		return
	}

	full, err := g.agent.handle(msg, func(token string) {
		adapter.send(AgentReply{
			Type: "delta", ID: msg.ID, SessionID: msg.SessionID,
			Channel: msg.Channel, Content: token,
		})
	})
	if err != nil {
		adapter.send(AgentReply{
			Type: "error", ID: msg.ID, SessionID: msg.SessionID,
			Channel: msg.Channel, Content: err.Error(),
		})
		return
	}
	adapter.send(AgentReply{
		Type: "reply", ID: msg.ID, SessionID: msg.SessionID,
		Channel: msg.Channel, Content: full,
	})
}

func (g *Gateway) start() error {
	for _, a := range g.adapters {
		if err := a.start(); err != nil {
			return err
		}
	}
	return nil
}

// ── Router ────────────────────────────────────────────────────────────────────

func resolveSessionID(channel, clientSessionID string) string {
	if channel == "cli" {
		return "cli"
	}
	if clientSessionID != "" {
		return clientSessionID
	}
	return fmt.Sprintf("web-%d", time.Now().UnixMilli())
}

// ── History serialisation helper ──────────────────────────────────────────────

func marshalHistory(history []HistoryEntry) string {
	b, _ := json.Marshal(history)
	return string(b)
}

File: 08-memory-rag/nodejs/src/gateway/gateway.ts
import type { ChannelAdapter } from '../channels/types.ts';
import type { ACPMessage, AgentReply } from './types.ts';
import type { Agent } from '../agent.ts';
import { resolveSessionId } from './router.ts';
import type { DB } from '../db.ts';

export class Gateway {
  private adapters = new Map<string, ChannelAdapter>();
  private agent: Agent;
  private db: DB | null;

  constructor(agent: Agent, db: DB | null = null) {
    this.agent = agent;
    this.db = db;
  }

  register(adapter: ChannelAdapter): void {
    this.adapters.set(adapter.name, adapter);
    adapter.onMessage((raw) => this.dispatch(raw));
  }

  private async dispatch(raw: ACPMessage): Promise<void> {
    const msg: ACPMessage = {
      ...raw,
      sessionId: resolveSessionId(raw.channel, raw.sessionId),
    };
    const adapter = this.adapters.get(msg.channel)!;

    // Handle reconnect: restore history and optionally resume execution.
    if ((raw as any).type === 'reconnect') {
      await this.handleReconnect(msg.sessionId, adapter);
      return;
    }

    try {
      const full = await this.agent.handle(msg, (token) => {
        adapter.send({ type: 'delta', id: msg.id, sessionId: msg.sessionId, channel: msg.channel, content: token });
      });
      adapter.send({ type: 'reply', id: msg.id, sessionId: msg.sessionId, channel: msg.channel, content: full });
    } catch (err: any) {
      adapter.send({ type: 'error', id: msg.id, sessionId: msg.sessionId, channel: msg.channel, content: err.message });
    }
  }

  private async handleReconnect(sessionId: string, adapter: ChannelAdapter): Promise<void> {
    if (!this.db) return;

    const status = this.db.getStatus(sessionId);

    // Always send history (empty array for new sessions) so the CLI/web knows to start prompting.
    const history = status ? this.db.loadHistory(sessionId) : [];
    adapter.send({
      type: 'history',
      id: '',
      sessionId,
      channel: adapter.name,
      content: JSON.stringify(history),
    });

    if (!status) return; // brand-new session — nothing more to do

    // Running/Paused sessions: resume execution with an injected prompt.
    // Success/Failed: read-only; do not trigger the LLM.
    if (status === 'Running' || status === 'Paused') {
      const resumeMsg: ACPMessage = {
        id: crypto.randomUUID(),
        sessionId,
        channel: adapter.name,
        content:
          '[System: 之前由于不可抗力中断，请根据以下历史继续执行，不要重新从头开始。如果有未完成的工具调用，请重新发起。]',
        timestamp: Date.now(),
      };
      try {
        const full = await this.agent.handle(resumeMsg, (token) => {
          adapter.send({ type: 'delta', id: resumeMsg.id, sessionId, channel: adapter.name, content: token });
        });
        adapter.send({ type: 'reply', id: resumeMsg.id, sessionId, channel: adapter.name, content: full });
      } catch (err: any) {
        adapter.send({ type: 'error', id: '', sessionId, channel: adapter.name, content: err.message });
      }
    }
  }

  async start(): Promise<void> {
    await Promise.all([...this.adapters.values()].map(a => a.start()));
  }
}
